contracts/contracts/lib/cometbft/CometbftLightClient.sol

```diff
@@ -0,0 +1,200 @@
+            commitSig = header.commit.signatures[i];
+            // no need to verify absent or nil votes.
+            if (commitSig.block_id_flag != TENDERMINTLIGHT_PROTO_GLOBAL_ENUMS.BlockIDFlag.BLOCK_ID_FLAG_COMMIT) {
+                continue;
+            }
+
+            (uint256 power, address validator) = ensureValidatorSubmission(i, commitSig.validator_address);
+
+            bytes memory message = voteSignBytesDelim(header.commit, LibSubnetActorStorage.appStorage().chainID, i);
+            bytes32 messageHash = sha256(message);
+
+            ensureValidSignature(messageHash, commitSig.signature, validator);
+
+            powerSoFar += power;
+        }
+
+        if(powerSoFar < (totalPower * 2 / 3)) {
+            revert NoQuorumFormed();
+        }
+    }
+
+    function checkCommitHash(SignedHeader.Data memory header) internal pure {
+        bytes32 expected = header.hash();
+        bytes32 actual = toBytes32(header.commit.block_id.hash);
+        if (actual != expected) revert InvalidCommitHash(expected, actual);
+    }
+
+    function ensureValidSignature(bytes32 messageHash, bytes memory signature, address validator) internal pure {
+        (address recovered, ECDSA.RecoverError ecdsaErr) = verify(messageHash, signature, validator);
+
+        if (ecdsaErr != ECDSA.RecoverError.NoError) {
+            revert InvalidSignature(messageHash, signature, validator, ecdsaErr);
+        }
+        if (recovered != validator) {
+            revert NotSigner(messageHash, signature, recovered, validator);
+        }
+    }
+
+    function toBytes32(bytes memory bz) internal pure returns (bytes32 ret) {
+        if (bz.length != 32) {
+            revert InvalidLength("bytes32", 32, bz.length);
+        }
+        assembly {
+            ret := mload(add(bz, 32))
+        }
+    }
+
+    function ensureValidatorSubmission(uint256 i, bytes memory incomingValidator) internal view returns (uint256 power, address validator) {
+        validator = LibPower.getActiveValidatorAddressByIndex(uint16(i));
+        ValidatorInfo memory info = LibPower.getActiveValidatorInfo(validator);
+
+        bytes20 expectedCometbftAccountId = toCometBFTAddress(info.metadata);
+
+        bytes20 incoming;
+        assembly {
+            // mload(add(b, 32)) loads the first 32 bytes of the actual data (after skipping the 32-byte length prefix).
+            // first 12 bytes discarded due to bytes20.
+            incoming := mload(add(incomingValidator, 32))
+        }
+
+        if (incoming != expectedCometbftAccountId) {
+            revert CometbftSignerNotValidator(expectedCometbftAccountId, incoming);
+        }
+
+        power = info.currentPower;
+    }
+
+    function voteSignBytesDelim(
+        Commit.Data memory commit,
+        string memory _chainID,
+        uint256 idx
+    ) internal pure returns (bytes memory) {
+        CanonicalVote.Data memory vote = CanonicalVote.Data({
```

Comment:
What prevents the signed header to contain an arbitrary `app_hash`? 
It doesn't seems to be used here.

Reply:
It's not clear to me what exactly `checkCommitHash` is checking, and how that relates specifically to `app_hash`. Please provide more docs.

Reply:
It still not clear what exactly `checkCommitHash` is checking. 

1. What `header.hash()` is including?
2. You wrote "This also makes sure the AppHash is not fabricated" without explaining how it's done.